In [2]:
# DisulfideBond Class Analysis Dictionary creation
# Author: Eric G. Suchanek, PhD.
# (c) 2023 Eric G. Suchanek, PhD., All Rights Reserved
# License: MIT
# Last Modification: 11/16/23
# Cα Cβ Sγ

import pandas as pd
import numpy as np

import pyvista as pv
from pyvista import set_plot_theme

import proteusPy
from proteusPy import *
from proteusPy.Disulfide import *
from proteusPy.DisulfideList import DisulfideList
from proteusPy.utility import print_memory_used
from proteusPy.DisulfideClasses import plot_count_vs_classid, plot_count_vs_class_df, enumerate_sixclass_fromlist

# pyvista setup for notebooks
pv.set_jupyter_backend('trame')

print_memory_used()

THEME = 'light'
PDB_BLOB = "https://github.com/suchanek/proteusPy/blob/0adcd3185604f87b5f366232473a961fb67e6ac9/proteusPy/data/PDB_SS_ALL_LOADER.pkl"


proteusPy 0.92.2: Memory Used: 0.23 GB


In [3]:

PDB_SS = Load_PDB_SS(verbose=True, subset=False)
PDB_SS.describe()


-> load_PDB_SS(): Reading /Users/egs/repos/proteusPy/proteusPy/data/PDB_SS_ALL_LOADER.pkl... 
-> load_PDB_SS(): Done reading /Users/egs/repos/proteusPy/proteusPy/data/PDB_SS_ALL_LOADER.pkl... 
    =========== RCSB Disulfide Database Summary ==============
       =========== Built: 2024-02-08 16:06:42 ==============
PDB IDs present:                    35818
Disulfides loaded:                  120494
Average structure resolution:       2.34 Å
Lowest Energy Disulfide:            2q7q_75D_140D
Highest Energy Disulfide:           1toz_456A_467A
Cα distance cutoff:                 8.00 Å
Total RAM Used:                     31.05 GB.
    ================= proteusPy: 0.89 =======================


In [4]:
sslist = PDB_SS.tclass.sslist_from_classid('02020')
classes = PDB_SS.tclass.classdict
classes

{'-LHSpiral': array(['2qhe_27A_126A', '2qhe_44A_105A', '2qhe_51A_98A', ...,
        '1j7z_58B_110B', '1d9i_42A_58A', '6do1_18A_373A'], dtype=object),
 '00002': array(['2qhe_84A_96A', '4nz3_185A_200A', '4nzr_142H_208H', ...,
        '1zgv_168A_182A', '1d9i_168A_182A', '6hyg_41A_101A'], dtype=object),
 '-LHHook': array(['3l4o_38C_86C', '3l75_144E_160E', '3ulv_95A_122A', ...,
        '6qwn_103A_110A', '4k3i_38C_86C', '5voc_647A_650A'], dtype=object),
 '00022': array(['2hew_70F_163F', '5hdz_216A_420A', '5fji_73A_90A', ...,
        '1j7m_15A_41A', '4unz_52A_277A', '6tjw_43A_270A'], dtype=object),
 '-RHStaple': array(['1j5h_37A_47A', '4lb7_2D_30D', '2hew_98F_183F', ...,
        '6tjw_5A_137B', '3l5f_330A_380A', '1tie_132A_139A'], dtype=object),
 '00202': array(['1zed_467A_474A', '2zxt_396A_398A', '2zwl_17L_22L',
        '7wg6_391B_525B', '6zi8_6A_11A', '7ezc_259A_262A',
        '5sk4_498A_502A', '8dt3_391C_525C', '6qn7_144H_215L',
        '2cl2_155A_165A', '2d2n_121A_126C', '5ccu_232A_237A',

In [ ]:

pbar = enumerate(classes)
sslist = DisulfideList([], 'tmp')

for idx, class_ss_list in classes.items():
    #class_ss_list = PDB_SS.tclass.sslist_from_classid(cls)
    print(f'idx: {idx} cls: {class_ss_list} len: {len(class_ss_list)}')
    #sslist = [PDB_SS[id] for id in class_ss_list]
    print(sslist)




: 

In [ ]:
PDB_SS.plot_binary_to_sixclass_incidence(light=True, save=False, savedir='/Users/egs/Documents/proteusPy')

In [ ]:
# the average structures for each class, returned as a 
# list of disulfides
cons_list = PDB_SS.tclass.load_consensus_file()
PDB_SS.plot_classes_vs_cutoff(.5, 50)

In [ ]:
sixclass_df = PDB_SS.tclass.sixclass_df.copy()
sixclass_df.sort_values(by='percentage', ascending=False, inplace=True)
sixclass_df.head()


In [ ]:
PDB_SS.tclass.classdf

In [ ]:
plot_count_vs_classid(PDB_SS.tclass.classdf, title='Disulfides by Binary Class')


In [ ]:
plot_count_vs_classid(PDB_SS.tclass.sixclass_df, title='Disulfides by Sextant Class')

# Disulfide Binary Class Breakdown

|   class_id | SS_Classname   | FXN        |   count |   incidence |
|-----------:|:---------------|:-----------|--------:|------------:|
|      00000 | -LHSpiral      | UNK        |   31513 |  0.261092   |
|      00002 | 00002          | UNK        |    5805 |  0.0480956  |
|      00020 | -LHHook        | UNK        |    3413 |  0.0282774  |
|      00022 | 00022          | UNK        |    1940 |  0.0160733  |
|      00200 | -RHStaple      | Allosteric |   12735 |  0.105512   |
|      00202 | 00202          | UNK        |     993 |  0.00822721 |
|      00220 | 00220          | UNK        |    5674 |  0.0470103  |
|      00222 | 00222          | UNK        |    5092 |  0.0421883  |
|      02000 | 02000          | UNK        |    4749 |  0.0393465  |
|      02002 | 02002          | UNK        |    3774 |  0.0312684  |
|      02020 | -LHStaple      | UNK        |    1494 |  0.0123781  |
|      02022 | 02022          | UNK        |     591 |  0.00489656 |
|      02200 | -RHHook        | UNK        |    5090 |  0.0421717  |
|      02202 | 02202          | UNK        |     533 |  0.00441602 |
|      02220 | -RHSpiral      | UNK        |    6751 |  0.0559335  |
|      02222 | 02222          | UNK        |    3474 |  0.0287828  |
|      20000 | ±LHSpiral      | UNK        |    3847 |  0.0318732  |
|      20002 | +LHSpiral      | UNK        |     875 |  0.00724956 |
|      20020 | ±LHHook        | UNK        |     803 |  0.00665302 |
|      20022 | +LHHook        | UNK        |     602 |  0.0049877  |
|      20200 | ±RHStaple      | UNK        |     419 |  0.0034715  |
|      20202 | +RHStaple      | UNK        |     293 |  0.00242757 |
|      20220 | ±RHHook        | Catalytic  |    1435 |  0.0118893  |
|      20222 | 20222          | UNK        |     488 |  0.00404318 |
|      22000 | -/+LHHook      | UNK        |    2455 |  0.0203402  |
|      22002 | 22002          | UNK        |    1027 |  0.00850891 |
|      22020 | ±LHStaple      | UNK        |    1046 |  0.00866633 |
|      22022 | +LHStaple      | UNK        |     300 |  0.00248556 |
|      22200 | -/+RHHook      | UNK        |    6684 |  0.0553783  |
|      22202 | +RHHook        | UNK        |     593 |  0.00491313 |
|      22220 | ±RHSpiral      | UNK        |    2544 |  0.0210776  |
|      22222 | +RHSpiral      | UNK        |    3665 |  0.0303653  |


In [ ]:
PDB_SS.plot_classes_vs_cutoff(.2, 50)

In [ ]:
PDB_SS.tclass.sslist_from_classid('+RHHook')

In [ ]:
# Remapping binary classes into six-space
# the tclass.binary_to_six_classes() function returns
# all possible combinations as a list of sextant-based
# strings. These variable names correspond to the names given by
# Hogg et al.

# most prevelent
LHSpiral_neg = PDB_SS.tclass.binary_to_six_class("00000")
RHSpiral_neg = PDB_SS.tclass.binary_to_six_class("02220")

LHSpiral_pminus = PDB_SS.tclass.binary_to_six_class("20000")
LHSpiral_plus = PDB_SS.tclass.binary_to_six_class("20002")

LHHook_neg = PDB_SS.tclass.binary_to_six_class("00020")
LHHook_pminus = PDB_SS.tclass.binary_to_six_class("20020")
LHHook_plus = PDB_SS.tclass.binary_to_six_class("20022")
LHHook_minus_plus = PDB_SS.tclass.binary_to_six_class("22000")

# Catalytic
RHHook_pminus = PDB_SS.tclass.binary_to_six_class("20220")

RHHook_minus = PDB_SS.tclass.binary_to_six_class("02200")

# Allosteric
RHStaple_neg = PDB_SS.tclass.binary_to_six_class("00200")

RHStaple_pminus = PDB_SS.tclass.binary_to_six_class("20200")
RHStaple_plus = PDB_SS.tclass.binary_to_six_class("20202")

LHStaple_neg = PDB_SS.tclass.binary_to_six_class("02020")
LHStaple_pminus = PDB_SS.tclass.binary_to_six_class("22020")
LHStaple_plus = PDB_SS.tclass.binary_to_six_class("22022")

RHSpiral_pminus = PDB_SS.tclass.binary_to_six_class("22220")
RHSpiral_plus = PDB_SS.tclass.binary_to_six_class("22222")



In [ ]:
PDB_SS.tclass.sixclass_df

# Examining a Catalytic Disulfide Class
RHHook_pminus is considered to be 'catalytic', so we can
look more closely in six-space by examining the number of disulfides within
each of the possible six-space strings computed above:

For each six-class ID in the combo list, return the sslist represented by
that class and calculate the number of SS in it. 
Plot the results.

In [ ]:
df = enumerate_sixclass_fromlist(RHHook_pminus)
plot_count_vs_class_df(df, 'RHH_pminus', theme='light')

In [ ]:
df2 = enumerate_sixclass_fromlist(RHStaple_neg)
plot_count_vs_class_df(df2, 'RHStaple_neg (Allosteric)', theme='light')

In [ ]:
df = enumerate_sixclass_fromlist(LHSpiral_neg)
plot_count_vs_class_df(df, title='LHSpiral_neg', theme='light')



In [ ]:
df4 = enumerate_sixclass_fromlist(RHSpiral_plus)
plot_count_vs_classid(df4, title='RHSpiral_plus', theme=THEME)

In [ ]:
from scipy.optimize import minimize
import numpy as np

def energy_function(x):
    chi1, chi2, chi3, chi4, chi5 = x
    energy = 2.0 * (np.cos(np.deg2rad(3.0 * chi1)) + np.cos(np.deg2rad(3.0 * chi5)))
    energy += np.cos(np.deg2rad(3.0 * chi2)) + np.cos(np.deg2rad(3.0 * chi4))
    energy += 3.5 * np.cos(np.deg2rad(2.0 * chi3)) + 0.6 * np.cos(np.deg2rad(3.0 * chi3)) + 10.1
    return energy

initial_guess = [-60.0, -60.0, 90.0, -60.0, -60.0] # initial guess for chi1, chi2, chi3, chi4, chi5
result = minimize(energy_function, initial_guess, method="Nelder-Mead")
minimum_energy = result.fun
inputs = result.x
inputs[0]


In [ ]:
best_ss = Disulfide('BestDisulfide')
best_ss.build_model(inputs[0], inputs[1], inputs[2], inputs[3], inputs[4])
best_ss.pprint()


In [ ]:
def download_and_save_binary(url, filename):
    import requests
    response = requests.get(url)
    content = response.content
    open(filename, "wb+").write(content)

# download_and_save_binary("https://github.com/suchanek/proteusPy/blob/0adcd3185604f87b5f366232473a961fb67e6ac9/proteusPy/data/PDB_SS_ALL_LOADER.pkl", "SS.pkl")